fem/gslib.cpp

```diff
@@ -853,6 +1915,247 @@ void FindPointsGSLIB::MapRefPosAndElemIndices()
+      sarray_transfer(evalSrcPt_t, &src, proc, 1, cr);
+   }
+
+   //evaluate points that are already local
+   {
+      Vector interp_vals(nlocal*ncomp);
+      interp_vals.UseDevice(true);
+
+      if (dim == 2)
+      {
+         InterpolateLocal2(field_in_evec,
+                           gsl_elem_temp,
+                           gsl_ref_temp,
+                           interp_vals,
+                           nlocal, ncomp,
+                           nel, dof1Dsol);
+      }
+      else
+      {
+         InterpolateLocal3(field_in_evec,
+                           gsl_elem_temp,
+                           gsl_ref_temp,
+                           interp_vals,
+                           nlocal, ncomp,
+                           nel, dof1Dsol);
+
+      }
+      MPI_Barrier(gsl_comm->c);
+
+      interp_vals.HostRead();
+
+      // now put these in correct positions
+      int interp_Offset = interp_vals.Size()/ncomp;
+      for (int i = 0; i < ncomp; i++)
+      {
+         for (int j = 0; j < nlocal; j++)
+         {
+            int pt_index = index_temp[j];
+            int idx = ordering == Ordering::byNODES ?
+                      pt_index + i*points_cnt :
+                      pt_index*ncomp + i;
+            field_out(idx) = AsConst(interp_vals)(j + interp_Offset*i);
+         }
+      }
+   }
+   MPI_Barrier(gsl_comm->c);
+
+   if (gsl_comm->np == 1)
+   {
+      array_free(&src);
+      return;
+   }
+
+   // evaluate points locally
+   {
+      int n = src.n;
+      const evalSrcPt_t *spt;
+      evalOutPt_t *opt;
+      spt = (evalSrcPt_t *)src.ptr;
+
+      // Copy to host vector
+      gsl_elem_temp.SetSize(n);
+      gsl_elem_temp.HostWrite();
+
+      gsl_ref_temp.SetSize(n*dim);
+      gsl_ref_temp.HostWrite();
+
+      spt = (evalSrcPt_t *)src.ptr;
+      //   opt = (evalOutPt_t *)outpt.ptr;
+      for (int i = 0; i < n; i++, ++spt)
+      {
+         gsl_elem_temp[i] = spt->el;
+         for (int d = 0; d < dim; d++)
+         {
+            gsl_ref_temp(i*dim + d) = spt->r[d];
+         }
+      }
+
+      Vector interp_vals(n*ncomp);
+      interp_vals.UseDevice(true);
+      if (dim == 2)
+      {
+         InterpolateLocal2(field_in_evec,
+                           gsl_elem_temp,
+                           gsl_ref_temp,
+                           interp_vals, n, ncomp,
+                           nel, dof1Dsol);
+      }
+      else
+      {
+         InterpolateLocal3(field_in_evec,
+                           gsl_elem_temp,
+                           gsl_ref_temp,
+                           interp_vals, n, ncomp,
+                           nel, dof1Dsol);
+      }
+      MPI_Barrier(gsl_comm->c);
+      interp_vals.HostRead();
+
+      // Now the interpolated values need to be sent back component wise
+      int Offset = interp_vals.Size()/ncomp;
+      for (int i = 0; i < ncomp; i++)
+      {
+         spt = (evalSrcPt_t *)src.ptr;
+         array_init(evalOutPt_t, &outpt, n);
+         outpt.n = n;
```

Comment:
I think this can be moved outside the i loop?

Reply:
The call to `sarray_transfer` redistributes the contents in parallel so we need to keep this inside the loop.